### PR TITLE
CODEBASE: Recheck all usages of typecasting with JSON.parse

### PR DESCRIPTION
**Pull request by catloversg**
While we search and fix lint errors, we (kind of) agree that we should not typecast the result of `JSON.parse` (e.g., `JSON.parse(saveString, Reviver) as Something`. This PR searches all references of `JSON.parse` and checks if they are typecasting in a questionable way. My guideline:
- It's fine to typecast to `unknown`.
- It's fine to typecast after or immediately before the validation code.
- Compromises:
  - If invalid data comes from a third-party tool (RFA client).
  - If proper validation requires an overcomplicated and risky refactor (the god object PlayerObject).

### CODE REVIEW

**Comment by catloversg** [comment]
In the latest commit, I refactored `loadStaneksGift`. There is a behavior change:
- Old: If `saveString` is an empty string, we load default data; otherwise, we parse `saveString` and pretend that everything is okay. If there is an error, the caller (`loadGame`) will handle it with a try-catch.
- New: Always pass `saveString` to JSON.parse in a try-catch. Validate the result. If it's invalid, we load default data, then show a dialog.

Rationale:
- BN13 was added in 793d9b34ce9848485216502ace7cda727267355e - v0.54.0 - 26 Sep 2021. `saveString` can only be invalid (empty string or invalid data) if the save file is older than v0.54.0 or it's a badly edited one.
- Proper validation.
- When the save string is invalid, instead of silently reset data, we show a dialog to warn the player. This is not a problem, unless they try to load an ancient/invalid save file.

From now on, I'll use this code for other "loader" functions if I need to change them (I'm testing another PR that changes `loadAllGangs`):
```js
let somethingData: unknown;
try {
  somethingData = JSON.parse(saveString, Reviver);
} catch (error) {
  console.error(error);
}
if (!validate(somethingData)) {
  console.error("Invalid somethingData:", saveString);
  somethingData = DefaultData;
  setTimeout(() => {
    dialogBoxCreate("Cannot load data of something. Something is reset.");
  }, 1000);
  return;
}
something = somethingData;
```

After #1774 is merged, with proper "loader" functions, we can gradually deal with legacy code in `loadGame` in a safe way.

File: src/CotMG/Helper.tsx [hunk 1]
```diff
@@ -6,11 +7,21 @@ import { StaneksGift } from "./StaneksGift";
...
 
 export function loadStaneksGift(saveString: string): void {
-  if (saveString) {
-    staneksGift = JSON.parse(saveString, Reviver) as StaneksGift;
-  } else {
+  let staneksGiftData: unknown;
+  try {
+    staneksGiftData = JSON.parse(saveString, Reviver);
+  } catch (error) {
+    console.error(error);
```
Comment (d0sboots) [inline src/CotMG/Helper.tsx:17]: Can we merge this condition into the other error handling? If someone edits their file incompetently, it would be good to be notified (more than just on the console)

File: src/SaveObject.ts [hunk 2]
```diff
@@ -220,11 +220,13 @@ class BitburnerSaveObject {
...
 
     if (!decodedSaveData || decodedSaveData === "") {
-      return Promise.reject(new Error("Save game is invalid"));
+      console.error("decodedSaveData:", decodedSaveData);
+      return Promise.reject(new Error("Save game is invalid. The save data cannot be decoded."));
     }
 
     let parsedSaveData;
     try {
+      // Typecasting here is fine. We will validate parsedSaveData immediately after this line.
```
Comment (d0sboots) [inline src/SaveObject.ts]: I don't think we actually need this cast, or it should be possible to get rid of with some code tweaks. The "if" block *does* do the same thing; TS should be able to see from the control-flow that the type is narrowed to this structure after this point. 

Doing it without the cast ensures we haven't missed an edge case.

File: test/jest/FullSave.test.ts [hunk 3]
```diff
@@ -11,12 +11,14 @@ import { Companies } from "../../src/Company/Companies";
 describe("Check Save File Continuity", () => {
   establishInitialConditions();
   // Calling getSaveString forces save info to update
-  saveObject.getSaveData();
+  saveObject.getSaveData().catch((error) => {
```
Comment (d0sboots) [inline test/jest/FullSave.test.ts]: Since this is a test, we want this error to propagate. Also, presumably we want this sequenced. So, `await` send like the answer here, but the test will have to become async.

**Review by d0sboots — COMMENTED** [review]
Note to self: Have to test this on the old saves too

File: test/jest/FullSave.test.ts (hunk 3 above)
Comment (catloversg) [reply test/jest/FullSave.test.ts]: I'll use `beforeAll`. Other solutions that do not work:
- async `describe`: `describe` does not accept an async function.
- If we put test code inside `then`, Jest does not recognize the tests.

**Comment by catloversg** [comment]
Unrelated question:
We have [`src\utils\helpers\typeAssertion.ts`](https://github.com/bitburner-official/bitburner-src/blob/75cf9c88b54479dffb0aa6146184ba337d92ec8e/src/utils/helpers/typeAssertion.ts) and [`src\utils\TypeAssertion.ts`](https://github.com/bitburner-official/bitburner-src/blob/75cf9c88b54479dffb0aa6146184ba337d92ec8e/src/utils/TypeAssertion.ts). Can I merge them later in another PR (`src\utils\helpers\typeAssertion.ts` -> `src\utils\TypeAssertion.ts`)?

**Comment by d0sboots** [comment]
> Unrelated question: We have [`src\utils\helpers\typeAssertion.ts`](https://github.com/bitburner-official/bitburner-src/blob/75cf9c88b54479dffb0aa6146184ba337d92ec8e/src/utils/helpers/typeAssertion.ts) and [`src\utils\TypeAssertion.ts`](https://github.com/bitburner-official/bitburner-src/blob/75cf9c88b54479dffb0aa6146184ba337d92ec8e/src/utils/TypeAssertion.ts). Can I merge them later in another PR (`src\utils\helpers\typeAssertion.ts` -> `src\utils\TypeAssertion.ts`)?

Please do, that's just silly. *Unless* there are circular dependency issues, but those should be solved by breaking individual functions out, not having two of the same file.

**Comment by d0sboots** [comment]
In testing this change with the old saves I have, I ran into this issue. I don't think this is a problem with this change directly, but rather an earlier undetected regression:

> Error: WorldDaemon is not a normal server. This is a bug. Please contact developers. (at "Loading Screen")

[test_save_0.47.2.json](https://github.com/user-attachments/files/17841573/test_save_0.47.2.json)

**Comment by d0sboots** [comment]
I confirmed that it has the same issue on `dev`. So, an uncaught regression. This PR seems fine across all the test saves I have.
